python/tvm/contrib/pipeline_executor.py

```diff
@@ -0,0 +1,559 @@
+            self.bindings = {}
+            self.io_owner = owner
+            self.binding_type = type_name
+
+        def get_binding_data_type(self, key):
+            """return binding data type"""
+            if isinstance(self.io_owner, PipelineConfig.ModuleWrapper):
+                return self.io_owner.get_data_type(key, self.binding_type)
+
+            return None
+
+        def __getitem__(self, key):
+            """return item by key"""
+            if key not in self.bindings:
+                data_type = self.get_binding_data_type(key)
+                if not data_type and isinstance(self.io_owner, PipelineConfig.ModuleWrapper):
+                    raise RuntimeError(f"Illegal name: {self.binding_type}:{key} cannot find")
+
+                self.bindings[key] = PipelineConfig.ModuleWrapper.Binding(
+                    self.io_owner, self.binding_type, key, data_type
+                )
+
+            return self.bindings[key]
+
+    def __init__(self):
+        self.last_mod_indx = 0
+        self.mod_wrapper = {}
+        self.input_bindings = self.BindingList(self, "input")
+        self.output_bindings = self.BindingList(self, "output")
+
+    def __str__(self):
+        """ Get configuration in string type"""
+        # Sort moudles
+        self.dag_topology_sort()
+
+        # get input
+        input_dump = "Inputs\n"
+        for input_name in self.input_bindings.bindings:
+            inf = self.input_bindings.bindings[input_name]
+            input_dump += inf.__repr__() + "\n"
+
+        # get connections
+        output = {}
+        connections_dump = "\nconnections\n"
+        for mod in self.mod_wrapper:
+            for _, interface in self.mod_wrapper[mod].output_bindings.bindings.items():
+                if interface.bindings:
+                    mname, dname = interface.get_name()
+                    iname = mname + ".output(" + dname + ")->"
+                    for dep in interface.bindings:
+                        dep_mname, dep_dname = dep.get_name()
+                        if isinstance(dep.io_owner, PipelineConfig.ModuleWrapper):
+                            iname += " " + dep_mname + "." + dep_dname
+                            connections_dump += "  |" + iname + "\n"
+                        else:
+                            output[dep_dname] = mname + ".output(" + dname + ")"
+
+        # get output
+        output_dump = "\noutput\n"
+        for name in sorted(output.keys()):
+            output_dump += "  |output(" + name + ") : " + output[name] + "\n"
+
+        return input_dump + output_dump + connections_dump
+
+    def __getitem__(self, key):
+        """return item by key"""
+        if isinstance(key, tvm.ir.module.IRModule):
+            if key not in self.mod_wrapper:
+                # self.last_mod_indx start from 1 and be initialize value,
+                # the final value for mod index would get generate by function
+                # dag_topology_sort
```

Comment:
In this case can we just let index unset until topological sort? It's weird to see module index has two meanings (adding order and topological order).

Reply:
fixed.